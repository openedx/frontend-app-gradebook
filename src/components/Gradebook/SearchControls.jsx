import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { Button, Icon, SearchField } from '@edx/paragon';

import {
  fetchGrades,
  fetchMatchingUserGrades,
} from '../../data/actions/grades';

/**
 * Controls for filtering the GradebookTable. Contains the "Edit Filters" button for opening the filter drawer
 * as well as the search box for searching by username/email.
 */
export class SearchControls extends React.Component {
  constructor(props) {
    super(props);
    this.onSubmit = this.onSubmit.bind(this);
    this.onChange = this.onChange.bind(this);
    this.onClear = this.onClear.bind(this);
  }

  /** Submitting searches for user matching the username/email in `value` */
  onSubmit(value) {
    this.props.searchForUser(
      this.props.courseId,
      value,
      this.props.selectedCohort,
      this.props.selectedTrack,
      this.props.selectedAssignmentType,
    );
  }

  /** Changing the search value stores the key in Gradebook. Currently unused */
  onChange(filterValue) {
    this.props.setFilterValue(filterValue);
  }

  /** Clearing the search box falls back to showing students with already applied filters */
  onClear() {
    this.props.getUserGrades(
      this.props.courseId,
      this.props.selectedCohort,
      this.props.selectedTrack,
      this.props.selectedAssignmentType,
    );
  }

  render() {
    return (
      <>
        <h4>Step 1: Filter the Grade Report</h4>
        <div className="d-flex justify-content-between">
          <Button
            id="edit-filters-btn"
            className="btn-primary align-self-start"
            onClick={this.props.toggleFilterDrawer}
          >
            <Icon className="fa fa-filter" /> Edit Filters
          </Button>
          <div>
            <SearchField
              onSubmit={this.onSubmit}
              inputLabel="Search for a learner"
              onChange={this.onChange}
              onClear={this.onClear}
              value={this.props.filterValue}
            />
            <small className="form-text text-muted search-help-text">Search by username, email, or student key</small>
          </div>
        </div>
      </>
    );
  }
}

SearchControls.defaultProps = {
  courseId: '',
  filterValue: '',
  selectedAssignmentType: '',
  selectedCohort: null,
  selectedTrack: null,
};

SearchControls.propTypes = {
  courseId: PropTypes.string,
  filterValue: PropTypes.string,
  setFilterValue: PropTypes.func.isRequired,
  toggleFilterDrawer: PropTypes.func.isRequired,
  // From Redux
  getUserGrades: PropTypes.func.isRequired,
  searchForUser: PropTypes.func.isRequired,
  selectedAssignmentType: PropTypes.string,
  selectedCohort: PropTypes.string,
  selectedTrack: PropTypes.string,
};

export const mapStateToProps = (state) => ({
  selectedAssignmentType: state.filters.assignmentType,
  selectedTrack: state.filters.track,
  selectedCohort: state.filters.cohort,
});

export const mapDispatchToProps = {
  getUserGrades: fetchGrades,
  searchForUser: fetchMatchingUserGrades,
};

export default connect(mapStateToProps, mapDispatchToProps)(SearchControls);
