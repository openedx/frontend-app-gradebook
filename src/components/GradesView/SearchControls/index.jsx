import React from 'react';

import { SearchField } from '@edx/paragon';
import useSearchControlsData from './hooks';

/**
 * Controls for filtering the GradebookTable. Contains the "Edit Filters" button for opening the filter drawer
 * as well as the search box for searching by username/email.
 */
export const SearchControls = () => {
  const {
    onSubmit,
    onBlur,
    onClear,
    searchValue,
    inputLabel,
    hintText,
  } = useSearchControlsData();

  return (
    <div>
      <SearchField
        onSubmit={onSubmit}
        inputLabel={inputLabel}
        onBlur={onBlur}
        onClear={onClear}
        value={searchValue}
      />
      <small className="form-text text-muted search-help-text">
        {hintText}
      </small>
    </div>
  );
};

SearchControls.propTypes = {};

export default SearchControls;
