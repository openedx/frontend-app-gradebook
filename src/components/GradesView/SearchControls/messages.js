import { defineMessages } from '@edx/frontend-platform/i18n';

const messages = defineMessages({
  label: {
    id: 'gradebook.GradesView.search.label',
    defaultMessage: 'Search for a learner',
    description: 'Text prompting a user to use this functionality to search for a learner',
  },
  hint: {
    id: 'gradebook.GradesView.search.hint',
    defaultMessage: 'Search by username, email, or student key',
    description: 'A hint explaining the ways a user can search',
  },
});

export default messages;
