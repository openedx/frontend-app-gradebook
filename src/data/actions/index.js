import { StrictDict } from '../../utils';

import app from './app';
import assignmentTypes from './assignmentTypes';
import cohorts from './cohorts';
import config from './config';
import filters from './filters';
import grades from './grades';
import roles from './roles';
import tracks from './tracks';

export default StrictDict({
  app,
  assignmentTypes,
  cohorts,
  config,
  filters,
  grades,
  roles,
  tracks,
});
