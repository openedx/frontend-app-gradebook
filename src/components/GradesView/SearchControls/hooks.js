import { useIntl } from '@edx/frontend-platform/i18n';

import { actions, selectors, thunkActions } from 'data/redux/hooks';

import messages from './messages';

/**
 * Controls for filtering the GradebookTable. Contains the "Edit Filters" button for opening the filter drawer
 * as well as the search box for searching by username/email.
 */
export const useSearchControlsData = () => {
  const { formatMessage } = useIntl();
  const searchValue = selectors.app.useSearchValue();
  const fetchGrades = thunkActions.grades.useFetchGrades();
  const setSearchValue = actions.app.useSetSearchValue();

  const onBlur = (e) => {
    setSearchValue(e.target.value);
  };

  const onClear = () => {
    setSearchValue('');
    fetchGrades();
  };

  const onSubmit = (newValue) => {
    setSearchValue(newValue);
    fetchGrades();
  };

  return {
    onSubmit,
    onBlur,
    onClear,
    searchValue,
    inputLabel: formatMessage(messages.label),
    hintText: formatMessage(messages.hint),
  };
};

export default useSearchControlsData;
