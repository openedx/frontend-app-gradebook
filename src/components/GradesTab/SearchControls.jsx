import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { Button, Icon, SearchField } from '@edx/paragon';
import { FormattedMessage } from '@edx/frontend-platform/i18n';

import actions from 'data/actions';
import selectors from 'data/selectors';
import thunkActions from 'data/thunkActions';

import messages from './messages';

/**
 * Controls for filtering the GradebookTable. Contains the "Edit Filters" button for opening the filter drawer
 * as well as the search box for searching by username/email.
 */
export class SearchControls extends React.Component {
  constructor(props) {
    super(props);
    this.onChange = this.onChange.bind(this);
    this.onClear = this.onClear.bind(this);
  }

  /** Changing the search value stores the key in Gradebook. Currently unused */
  onChange(searchValue) {
    this.props.setSearchValue(searchValue);
  }

  onClear() {
    this.props.setSearchValue('');
    this.props.fetchGrades();
  }

  render() {
    return (
      <>
        <h4><FormattedMessage {...messages.filterStepHeading} /></h4>
        <div className="d-flex justify-content-between">
          <Button
            id="edit-filters-btn"
            className="btn-primary align-self-start"
            onClick={this.props.toggleFilterDrawer}
          >
            <Icon className="fa fa-filter" /> <FormattedMessage {...messages.editFilters} />
          </Button>
          <div>
            <SearchField
              onSubmit={this.props.fetchGrades}
              inputLabel={<FormattedMessage {...messages.searchLabel} />}
              onChange={this.onChange}
              onClear={this.onClear}
              value={this.props.searchValue}
            />
            <small className="form-text text-muted search-help-text">
              <FormattedMessage {...messages.searchHint} />
            </small>
          </div>
        </div>
      </>
    );
  }
}

SearchControls.propTypes = {
  // From Redux
  fetchGrades: PropTypes.func.isRequired,
  searchValue: PropTypes.string.isRequired,
  setSearchValue: PropTypes.func.isRequired,
  toggleFilterDrawer: PropTypes.func.isRequired,
};

export const mapStateToProps = (state) => ({
  searchValue: selectors.app.searchValue(state),
});

export const mapDispatchToProps = {
  fetchGrades: thunkActions.grades.fetchGrades,
  setSearchValue: actions.app.setSearchValue,
  toggleFilterDrawer: thunkActions.app.filterMenu.toggle,
};

export default connect(mapStateToProps, mapDispatchToProps)(SearchControls);
