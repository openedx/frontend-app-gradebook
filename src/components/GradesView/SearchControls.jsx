import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { SearchField } from '@edx/paragon';
import { FormattedMessage } from '@edx/frontend-platform/i18n';

import actions from 'data/actions';
import selectors from 'data/selectors';
import thunkActions from 'data/thunkActions';

import messages from './SearchControls.messages';

/**
 * Controls for filtering the GradebookTable. Contains the "Edit Filters" button for opening the filter drawer
 * as well as the search box for searching by username/email.
 */
export class SearchControls extends React.Component {
  constructor(props) {
    super(props);
    this.onBlur = this.onBlur.bind(this);
    this.onClear = this.onClear.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  /** Changing the search value stores the key in Gradebook. Currently unused */
  onBlur(e) {
    const { value } = e.target;
    if (this.props.searchValue !== value) {
      this.props.setSearchValue(value);
    }
  }

  onClear() {
    this.props.setSearchValue('');
    this.props.fetchGrades();
  }

  onSubmit(value) {
    if (this.props.searchValue !== value) {
      this.props.setSearchValue(value);
    }
    this.props.fetchGrades();
  }

  render() {
    return (
      <div>
        <SearchField
          onSubmit={this.onSubmit}
          inputLabel={<FormattedMessage {...messages.label} />}
          onBlur={this.onBlur}
          onClear={this.onClear}
          value={this.props.searchValue}
        />
        <small className="form-text text-muted search-help-text">
          <FormattedMessage {...messages.hint} />
        </small>
      </div>
    );
  }
}

SearchControls.propTypes = {
  // From Redux
  fetchGrades: PropTypes.func.isRequired,
  searchValue: PropTypes.string.isRequired,
  setSearchValue: PropTypes.func.isRequired,
};

export const mapStateToProps = (state) => ({
  searchValue: selectors.app.searchValue(state),
});

export const mapDispatchToProps = {
  fetchGrades: thunkActions.grades.fetchGrades,
  setSearchValue: actions.app.setSearchValue,
};

export default connect(mapStateToProps, mapDispatchToProps)(SearchControls);
