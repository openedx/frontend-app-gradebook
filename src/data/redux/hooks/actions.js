import { StrictDict } from 'utils';
import actions from 'data/actions';
import { actionHook } from './utils';

const app = StrictDict({
  useSetLocalFilter: actionHook(actions.app.setLocalFilter),
  useSetSearchValue: actionHook(actions.app.setSearchValue),
  useSetShowImportSuccessToast: actionHook(actions.app.setShowImportSuccessToast),
  useSetView: actionHook(actions.app.setView),
  useCloseModal: actionHook(actions.app.closeModal),
  useSetModalState: actionHook(actions.app.setModalState),
});

const filters = StrictDict({
  useUpdateAssignment: actionHook(actions.filters.update.assignment),
  useUpdateAssignmentLimits: actionHook(actions.filters.update.assignmentLimits),
  useUpdateAssignmentType: actionHook(actions.filters.update.assignmentType),
  useUpdateCohort: actionHook(actions.filters.update.cohort),
  useUpdateCourseGradeLimits: actionHook(actions.filters.update.courseGradeLimits),
  useUpdateIncludeCourseRoleMembers: actionHook(actions.filters.update.includeCourseRoleMembers),
  useUpdateTrack: actionHook(actions.filters.update.track),
  useResetFilters: actionHook(actions.filters.reset),
});

const grades = StrictDict({
  useDoneViewingAssignment: actionHook(actions.grades.doneViewingAssignment),
  useDownloadBulkGradesReport: actionHook(actions.grades.downloadReport.bulkGrades),
  useDownloadInterventionReport: actionHook(actions.grades.downloadReport.intervention),
  useToggleGradeFormat: actionHook(actions.grades.toggleGradeFormat),
  useCloseBanner: actionHook(actions.grades.banner.close),
});

export default StrictDict({
  app,
  filters,
  grades,
});
