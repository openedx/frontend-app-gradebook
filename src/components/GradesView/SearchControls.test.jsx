import React from 'react';
import { shallow } from 'enzyme';

import selectors from 'data/selectors';
import actions from 'data/actions';
import thunkActions from 'data/thunkActions';
import { mapDispatchToProps, mapStateToProps, SearchControls } from './SearchControls';

jest.mock('@edx/paragon', () => ({
  Icon: 'Icon',
  Button: 'Button',
  SearchField: 'SearchField',
}));
jest.mock('data/selectors', () => ({
  __esModule: true,
  default: {
    app: {
      searchValue: jest.fn(state => ({ searchValue: state })),
    },
  },
}));
jest.mock('data/thunkActions', () => ({
  __esModule: true,
  default: {
    grades: {
      fetchGrades: jest.fn().mockName('thunkActions.grades.fetchGrades'),
    },
    app: {
      filterMenu: { toggle: jest.fn().mockName('thunkActions.app.filterMenu') },
    },
  },
}));

describe('SearchControls', () => {
  let props;

  beforeEach(() => {
    jest.resetAllMocks();
    props = {
      searchValue: 'alice',
      setSearchValue: jest.fn(),
      fetchGrades: jest.fn().mockName('fetchGrades'),
    };
  });

  const searchControls = (overriddenProps) => {
    props = { ...props, ...overriddenProps };
    return shallow(<SearchControls {...props} />);
  };

  describe('Component', () => {
    describe('Snapshots', () => {
      test('basic snapshot', () => {
        const wrapper = searchControls();
        wrapper.instance().onBlur = jest.fn().mockName('onBlur');
        wrapper.instance().onClear = jest.fn().mockName('onClear');
        wrapper.instance().onSubmit = jest.fn().mockName('onSubmit');
        expect(wrapper.instance().render()).toMatchSnapshot();
      });
    });

    describe('onBlur', () => {
      it('saves the changed search value to Gradebook state', () => {
        const wrapper = searchControls();
        wrapper.instance().onBlur({ target: { value: 'bob' } });
        expect(props.setSearchValue).toHaveBeenCalledWith('bob');
      });

      it('doesnt save the same as previous value to Gradebook state', () => {
        const wrapper = searchControls({ searchValue: 'bob' });
        wrapper.instance().onBlur({ target: { value: 'bob' } });
        expect(props.setSearchValue).not.toHaveBeenCalled();
      });
    });

    describe('onChange', () => {
      it('sets search value to empty string and calls fetchGrades', () => {
        const wrapper = searchControls();
        wrapper.instance().onClear();
        expect(props.setSearchValue).toHaveBeenCalledWith('');
        expect(props.fetchGrades).toHaveBeenCalled();
      });
    });

    describe('onSubmit', () => {
      it('saves the changed search value to Gradebook state and calls fetchGrades', () => {
        const wrapper = searchControls();
        wrapper.instance().onSubmit('bob');
        expect(props.setSearchValue).toHaveBeenCalledWith('bob');
        expect(props.fetchGrades).toHaveBeenCalled();
      });

      it('doesnt save the same as previous value to Gradebook state and calls fetchGrades', () => {
        const wrapper = searchControls({ searchValue: 'bob' });
        wrapper.instance().onSubmit('bob');
        expect(props.setSearchValue).not.toHaveBeenCalled();
        expect(props.fetchGrades).toHaveBeenCalled();
      });
    });

    describe('mapStateToProps', () => {
      const testState = { never: 'gonna', give: 'you up' };
      test('searchValue from app.searchValue', () => {
        expect(
          mapStateToProps(testState).searchValue,
        ).toEqual(selectors.app.searchValue(testState));
      });
    });
    describe('mapDispatchToProps', () => {
      test('fetchGrades from thunkActions.grades.fetchGrades', () => {
        expect(mapDispatchToProps.fetchGrades).toEqual(thunkActions.grades.fetchGrades);
      });

      test('setSearchValue from actions.app.setSearchValue', () => {
        expect(mapDispatchToProps.setSearchValue).toEqual(actions.app.setSearchValue);
      });
    });
  });
});
