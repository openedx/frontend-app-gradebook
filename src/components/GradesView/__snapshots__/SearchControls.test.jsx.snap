// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchControls Component Snapshots basic snapshot 1`] = `
<div>
  <SearchField
    inputLabel={
      <FormattedMessage
        defaultMessage="Search for a learner"
        description="Text prompting a user to use this functionality to search for a learner"
        id="gradebook.GradesView.search.label"
      />
    }
    onBlur={[MockFunction onBlur]}
    onClear={[MockFunction onClear]}
    onSubmit={[MockFunction onSubmit]}
    value="alice"
  />
  <small
    className="form-text text-muted search-help-text"
  >
    <FormattedMessage
      defaultMessage="Search by username, email, or student key"
      description="A hint explaining the ways a user can search"
      id="gradebook.GradesView.search.hint"
    />
  </small>
</div>
`;
